File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter14/FingerPaint/FingerPaint/MainPage.File.cpp
//
// MainPage.File.cpp
// Partial implementation of the MainPage class.
//

#include "pch.h"
#include "MainPage.xaml.h"

using namespace FingerPaint;

using namespace Concurrency;
using namespace Microsoft::WRL;
using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::Graphics::Imaging;
using namespace Windows::Storage;
using namespace Windows::Storage::Pickers;
using namespace Windows::Storage::Streams;
using namespace Windows::UI;
using namespace Windows::UI::Input;
using namespace Windows::UI::Popups;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Controls::Primitives;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Media;
using namespace Windows::UI::Xaml::Media::Imaging;
using namespace Windows::UI::Xaml::Navigation;

task<void> MainPage::CreateNewBitmapAndPixelArray()
{
    bitmap = ref new WriteableBitmap((int)this->ActualWidth, (int)this->ActualHeight);

    InitializeBitmap();

    // Set whole bitmap to white
    int length = 4 * bitmap->PixelWidth * bitmap->PixelHeight;
    for (int index = 0; index < length; index++)
        pixels[index] = 0xFF;

    // Invalidate the WriteableBitmap
    bitmap->Invalidate();

    // Set application settings
    appSettings->LoadedFilePath = nullptr;
    appSettings->LoadedFilename = nullptr;
    appSettings->IsImageModified = false;

    return task<void>([] () {});
}

task<void> MainPage::LoadBitmapFromFile()
{
    StorageFolder^ localFolder = ApplicationData::Current->LocalFolder;
    task<StorageFile^> getFileTask = create_task(localFolder->GetFileAsync("FingerPaint.png"));
    return getFileTask.then([this] (StorageFile^ storageFile)
    {
        task<void> loadBitmapTask = create_task(LoadBitmapFromFile(storageFile));
        loadBitmapTask.then([] ()
        {
        });
    });
}

task<void> MainPage::LoadBitmapFromFile(StorageFile^ storageFile)
{
    task<IRandomAccessStreamWithContentType^> openReadTask = create_task(storageFile->OpenReadAsync());
   
    return openReadTask.then([this](IRandomAccessStreamWithContentType^ stream)
    {
        task<BitmapDecoder^> createDecoderTask = create_task(BitmapDecoder::CreateAsync(stream));
        createDecoderTask.then([this] (BitmapDecoder^ decoder)
        {
            task<BitmapFrame^> getFrameTask = create_task(decoder->GetFrameAsync(0));
            getFrameTask.then([this](BitmapFrame^ bitmapFrame)
            {
                task<PixelDataProvider^> getPixelDataTask = 
                    create_task(bitmapFrame->GetPixelDataAsync(BitmapPixelFormat::Bgra8,
                                                               BitmapAlphaMode::Premultiplied,
                                                               ref new BitmapTransform(),
                                                               ExifOrientationMode::RespectExifOrientation,
                                                               ColorManagementMode::ColorManageToSRgb));
                getPixelDataTask.then([this, bitmapFrame](PixelDataProvider^ dataProvider)
                {
                    Array<byte>^ pSrcPixels = dataProvider->DetachPixelData();
                    bitmap = ref new WriteableBitmap((int)bitmapFrame->PixelWidth,
                                                     (int)bitmapFrame->PixelHeight);
                    InitializeBitmap();

                    for (unsigned int i = 0; i < pSrcPixels->Length; i++)
                        pixels[i] = pSrcPixels[i];
                });
            });
        });
    });
}

void MainPage::InitializeBitmap()
{
    // Obtain IBufferByteAccess
    IBuffer^ buffer = bitmap->PixelBuffer;
    ComPtr<IBufferByteAccess> pBufferByteAccess;
    ComPtr<IUnknown> pBuffer((IUnknown*)buffer);
    pBuffer.As(&pBufferByteAccess);

    // Get pointer to pixel bytes
    pBufferByteAccess->Buffer(&pixels);

    // Set to Image element
    image->Source = bitmap;
    CalculateImageScaleAndOffset();
}

void MainPage::OnAddAppBarButtonClick(Object^ sender, RoutedEventArgs^ args)
{
    Button^ button = dynamic_cast<Button^>(sender);
    button->IsEnabled = false;

    task<void> checkIfOkToTrashTask = 
        create_task(MainPage::CheckIfOkToTrashFile(&MainPage::CreateNewBitmapAndPixelArray));
    checkIfOkToTrashTask.then([this, button] ()
    {
        button->IsEnabled = true;
        this->BottomAppBar->IsOpen = false;
    }, task_continuation_context::use_current());
}

void MainPage::OnOpenAppBarButtonClick(Object^ sender, RoutedEventArgs^ args)
{
    Button^ button = dynamic_cast<Button^>(sender);
    button->IsEnabled = false;

    task<void> checkIfOkToTrashTask = 
        create_task(MainPage::CheckIfOkToTrashFile(&MainPage::LoadFileFromOpenPicker));
    checkIfOkToTrashTask.then([this, button] ()
    {
        button->IsEnabled = true;
        this->BottomAppBar->IsOpen = false;
    });
}

task<void> MainPage::CheckIfOkToTrashFile(task<void> (MainPage::*commandAction)(void))
{
    if (!appSettings->IsImageModified)
    {
        return (this->*commandAction)();
    }

    String^ message = "Do you want to save changes to " +
        (appSettings->LoadedFilePath == nullptr ? "(untitled)" : appSettings->LoadedFilePath) + "?";

    MessageDialog^ msgdlg = ref new MessageDialog(message, "Finger Paint");
    msgdlg->Commands->Append(ref new UICommand("Save", nullptr, "save"));
    msgdlg->Commands->Append(ref new UICommand("Don't save", nullptr, "dont"));
    msgdlg->Commands->Append(ref new UICommand("Cancel", nullptr, "cancel"));
    msgdlg->DefaultCommandIndex = 0;
    msgdlg->CancelCommandIndex = 2;

    task<IUICommand^> messageDialogShowTask(msgdlg->ShowAsync());
    return messageDialogShowTask.then([this, commandAction] (IUICommand^ command)
    {
        String^ strCommand = dynamic_cast<String^>(command->Id);

        if (strCommand == "cancel")
            return task<void>([]{});

        if (strCommand == "dont")
            return (this->*commandAction)();

        // Need to save file -- can do so directly with filename
        if (appSettings->LoadedFilePath != nullptr)
        {
            task<bool> saveToFileTask(MainPage::SaveWithErrorNotification(appSettings->LoadedFilePath));
            return saveToFileTask.then([this,commandAction] (bool success)
            {
                if (success)
                    return (this->*commandAction)();
                else
                    return task<void>([]{});
            });
        }

        // Otherwise need to get file from picker
        task<StorageFile^> getFileTask = create_task(MainPage::GetFileFromSavePicker());
        return getFileTask.then([this,commandAction](StorageFile^ storageFile)
        {
            if (storageFile == nullptr)
                return task<void>([]{});;

            task<bool> saveToFileTask(MainPage::SaveWithErrorNotification(storageFile));
            return saveToFileTask.then([this,commandAction] (bool success)
            {
                if (success)
                    return (this->*commandAction)();
                else
                    return task<void>([]{});
            });
        });
    });
}

task<void> MainPage::LoadFileFromOpenPicker()
{
    // Create FileOpenPicker
    FileOpenPicker^ picker = ref new FileOpenPicker();
    picker->SuggestedStartLocation = PickerLocationId::PicturesLibrary;

    // Initialize with filename extensions
    IVectorView<BitmapCodecInformation^>^ codecInfos = BitmapDecoder::GetDecoderInformationEnumerator();

    for each (BitmapCodecInformation^ codecInfo in codecInfos)
        for each (String^ extension in codecInfo->FileExtensions)
            picker->FileTypeFilter->Append(extension);

    // Get the selected file
    task<StorageFile^> pickFileTask = create_task(picker->PickSingleFileAsync());
    return pickFileTask.then([this] (StorageFile^ storageFile)
    {
        if (storageFile != nullptr)
        {
            task<void> loadBitmapTask = create_task(LoadBitmapFromFile(storageFile));
            loadBitmapTask.then([this, storageFile] (task<void> thisTask)
            {
                Exception^ exception = nullptr;

                try
                {
                    thisTask.get();
                }
                catch (Exception^ exc)
                {
                    exception = exc;
                }

                if (exception != nullptr)
                {
                    MessageDialog^ msgdlg = 
                        ref new MessageDialog("The image file could not be loaded. " +
                                              "The system reports an error of: " + exception,
                                              "Finger Paint"); 
                    task<IUICommand^> msgdlgShowTask = create_task(msgdlg->ShowAsync());
                    msgdlgShowTask.then([](IUICommand^)
                    {
                        return;
                    });
                 }
                 appSettings->LoadedFilePath = storageFile->Path;
                 appSettings->LoadedFilename = storageFile->Name;
                 appSettings->IsImageModified = false; 
            });
        }
    });
}

void MainPage::OnSaveAppBarButtonClick(Object^ sender, RoutedEventArgs^ args)
{
    Button^ button = dynamic_cast<Button^>(sender);
    button->IsEnabled = false;

    if (appSettings->LoadedFilePath != nullptr)
    {
        task<bool> saveFileTask = create_task(SaveWithErrorNotification(appSettings->LoadedFilePath));
        saveFileTask.then([button] (bool)
        {
            button->IsEnabled = true;
        });
    }
    else
    {
        task<StorageFile^> getFileTask = create_task(GetFileFromSavePicker());
        getFileTask.then([this, button] (StorageFile^ storageFile)
        {
            if (storageFile != nullptr)
            {
                task<bool> saveFileTask = create_task(SaveWithErrorNotification(storageFile));
                saveFileTask.then([button] (bool) 
                {
                    button->IsEnabled = true;
                });
            }
        });
    }
}

void MainPage::OnSaveAsAppBarButtonClick(Object^ sender, RoutedEventArgs^ args)
{
    task<StorageFile^> getFileTask = create_task(GetFileFromSavePicker());
    getFileTask.then([this] (StorageFile^ storageFile)
    {
        if (storageFile != nullptr)
        {
            task<bool> saveFileTask = create_task(SaveWithErrorNotification(storageFile));
            saveFileTask.then([] (bool) 
            {
            });
        }
    });
}

task<StorageFile^> MainPage::GetFileFromSavePicker()
{
    FileSavePicker^ picker = ref new FileSavePicker();
    picker->SuggestedStartLocation = PickerLocationId::PicturesLibrary;
    picker->SuggestedFileName = appSettings->LoadedFilename == nullptr ? "MyFingerPainting" : appSettings->LoadedFilename;

    // Get the encoder information
    Map<String^, Guid>^ imageTypes = ref new Map<String^, Guid>();
    IVectorView<BitmapCodecInformation^>^ codecInfos = 
                        BitmapEncoder::GetEncoderInformationEnumerator();

    for each (BitmapCodecInformation^ codecInfo in codecInfos)
    {
        Vector<String^>^ extensions = ref new Vector<String^>();

        for each (String^ extension in codecInfo->FileExtensions)
            extensions->Append(extension);

        wchar_t filetype[100];
        wcscpy_s(filetype, 100, codecInfo->FriendlyName->Data());
        int spaceIndex = wcscspn(filetype, L" ");
        filetype[spaceIndex] = '\0';
        picker->FileTypeChoices->Insert(ref new String(filetype), extensions);

        for each (String^ mimeType in codecInfo->MimeTypes)
            imageTypes->Insert(mimeType, codecInfo->CodecId);
    }

    task<StorageFile^> pickFileTask = create_task(picker->PickSaveFileAsync());
    return pickFileTask;
}

task<bool> MainPage::SaveWithErrorNotification(String^ filename)
{
    task<StorageFile^> getFileTask = create_task(StorageFile::GetFileFromPathAsync(filename));
    return getFileTask.then([this] (StorageFile^ storageFile)
    {
        task<bool> saveFileTask = create_task(SaveWithErrorNotification(storageFile));
        return saveFileTask.then([] (bool success)
        {
            return success;
        });
    });
}

task<bool> MainPage::SaveWithErrorNotification(StorageFile^ storageFile)
{
    task<void> saveBitmapTask = create_task(SaveBitmapToFile(storageFile));
    return saveBitmapTask.then([this, storageFile] (task<void> thisTask)
    {
        String^ exception = nullptr;

        try
        {
            thisTask.get();
        }
        catch (Exception^ exc)
        {
            exception = exc->Message;
        }

        if (exception != nullptr)
        {
            MessageDialog^ msgdlg =
                ref new MessageDialog("The image file could not be saved. " +
                                      "The system reports an error of: " + exception,
                                      "Finger Paint");
            task<IUICommand^> msgdlgShowTask = create_task(msgdlg->ShowAsync());
            msgdlgShowTask.then([] (IUICommand^)
            {
                return false;
            });
        }

        appSettings->LoadedFilename = storageFile->Name;
        appSettings->LoadedFilePath = storageFile->Path;
        appSettings->IsImageModified = false;
        return true;
    });
}

task<void> MainPage::SaveBitmapToFile()
{
    StorageFolder^ localFolder = ApplicationData::Current->LocalFolder;
    task<StorageFile^> createFileTask = create_task(localFolder->CreateFileAsync("FingerPaint.png",
                                                                                 CreationCollisionOption::ReplaceExisting));
    return createFileTask.then([this](StorageFile^ storageFile)
    {
        task<void> saveBitmapTask = create_task(SaveBitmapToFile(storageFile));
        return saveBitmapTask.then([] ()
        {
        });
    });
}

task<void> MainPage::SaveBitmapToFile(String^ filename)
{
    task<StorageFile^> getFileTask = create_task(StorageFile::GetFileFromPathAsync(filename));
    return getFileTask.then([this] (StorageFile^ storageFile)
    {
        task<void> saveBitmapTask = create_task(SaveBitmapToFile(storageFile));
        return saveBitmapTask.then([] ()
        {
        });
    });
}

task<void> MainPage::SaveBitmapToFile(StorageFile^ storageFile)
{
    task<IRandomAccessStream^> openFileTask = create_task(storageFile->OpenAsync(FileAccessMode::ReadWrite));
    return openFileTask.then([this] (IRandomAccessStream^ stream)
    {
        task<BitmapEncoder^> createEncoderTask = create_task(BitmapEncoder::CreateAsync(BitmapEncoder::PngEncoderId, stream));
        return createEncoderTask.then([this] (BitmapEncoder^ encoder)
        {
            unsigned int arraySize = 4 * bitmap->PixelWidth * bitmap->PixelHeight;
            Array<byte>^ pixelArray = ref new Array<byte>(pixels, arraySize);

            encoder->SetPixelData(BitmapPixelFormat::Bgra8,
                                  BitmapAlphaMode::Premultiplied,
                                  (unsigned int)bitmap->PixelWidth,
                                  (unsigned int)bitmap->PixelHeight,
                                  96, 96,
                                  pixelArray); // ref new Array<byte>(pixels, arraySize));

            task<void> flushTask = create_task(encoder->FlushAsync());
            return flushTask.then([] ()
            {
                return;
            });
        });
    });
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter17/FingerPaint/FingerPaint/MainPage.File.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Threading.Tasks;
using Windows.Graphics.Imaging;
using Windows.Storage;
using Windows.Storage.Pickers;
using Windows.Storage.Streams;
using Windows.UI.Popups;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media.Imaging;

namespace FingerPaint
{
    public sealed partial class MainPage : Page
    {
        WriteableBitmap bitmap;
        Stream pixelStream;
        byte[] pixels;

        async Task CreateNewBitmapAndPixelArray()
        {
            bitmap = new WriteableBitmap((int)this.ActualWidth, (int)this.ActualHeight);
            pixels = new byte[4 * bitmap.PixelWidth * bitmap.PixelHeight];

            // Set whole bitmap to white
            for (int index = 0; index < pixels.Length; index++)
                pixels[index] = 0xFF;

            await InitializeBitmap();

            appSettings.LoadedFilePath = null;
            appSettings.LoadedFilename = null;
            appSettings.IsImageModified = false;
        }

        async Task LoadBitmapFromFile(StorageFile storageFile)
        {
            using (IRandomAccessStreamWithContentType stream = await storageFile.OpenReadAsync())
            {
                BitmapDecoder decoder = await BitmapDecoder.CreateAsync(stream);
                BitmapFrame bitmapframe = await decoder.GetFrameAsync(0);
                PixelDataProvider dataProvider =
                    await bitmapframe.GetPixelDataAsync(BitmapPixelFormat.Bgra8,
                                                        BitmapAlphaMode.Premultiplied,
                                                        new BitmapTransform(),
                                                        ExifOrientationMode.RespectExifOrientation,
                                                        ColorManagementMode.ColorManageToSRgb);
                pixels = dataProvider.DetachPixelData();
                bitmap = new WriteableBitmap((int)bitmapframe.PixelWidth,
                                             (int)bitmapframe.PixelHeight);
                await InitializeBitmap();
            }
        }

        async Task InitializeBitmap()
        {
            pixelStream = bitmap.PixelBuffer.AsStream();
            await pixelStream.WriteAsync(pixels, 0, pixels.Length);
            bitmap.Invalidate();
            image.Source = bitmap;
            CalculateImageScaleAndOffset();
        }

        async void OnAddAppBarButtonClick(object sender, RoutedEventArgs args)
        {
            Button button = sender as Button;
            button.IsEnabled = false;

            await CheckIfOkToTrashFile(CreateNewBitmapAndPixelArray);

            button.IsEnabled = true;
            this.BottomAppBar.IsOpen = false;
        }

        async void OnOpenAppBarButtonClick(object sender, RoutedEventArgs args)
        {
            Button button = sender as Button;
            button.IsEnabled = false;

            await CheckIfOkToTrashFile(LoadFileFromOpenPicker);

            button.IsEnabled = true;
            this.BottomAppBar.IsOpen = false;
        }

        async Task CheckIfOkToTrashFile(Func<Task> commandAction)
        {
            if (!appSettings.IsImageModified)
            {
                await commandAction();
                return;
            }

            string message =
                String.Format("Do you want to save changes to {0}?",
                              appSettings.LoadedFilePath ?? "(untitled)");

            MessageDialog msgdlg = new MessageDialog(message, "Finger Paint");
            msgdlg.Commands.Add(new UICommand("Save", null, "save"));
            msgdlg.Commands.Add(new UICommand("Don't Save", null, "dont"));
            msgdlg.Commands.Add(new UICommand("Cancel", null, "cancel"));
            msgdlg.DefaultCommandIndex = 0;
            msgdlg.CancelCommandIndex = 2;
            IUICommand command = await msgdlg.ShowAsync();

            if ((string)command.Id == "cancel")
                return;

            if ((string)command.Id == "dont")
            {
                await commandAction();
                return;
            }

            if (appSettings.LoadedFilePath == null)
            {
                StorageFile storageFile = await GetFileFromSavePicker();

                if (storageFile == null)
                    return;

                appSettings.LoadedFilePath = storageFile.Path;
                appSettings.LoadedFilename = storageFile.Name;
            }

            string exception = null;

            try
            {
                await SaveBitmapToFile(appSettings.LoadedFilePath);
            }
            catch (Exception exc)
            {
                exception = exc.Message;
            }

            if (exception != null)
            {
                msgdlg = new MessageDialog("The image file could not be saved. " +
                                           "The system reports an error of: " + exception,
                                           "Finger Paint");
                await msgdlg.ShowAsync();
                return;
            }

            await commandAction();
        }

        async Task LoadFileFromOpenPicker()
        {
            // Create FileOpenPicker
            FileOpenPicker picker = new FileOpenPicker();
            picker.SuggestedStartLocation = PickerLocationId.PicturesLibrary;

            // Initialize with filename extensions
            IReadOnlyList<BitmapCodecInformation> codecInfos =
                                    BitmapDecoder.GetDecoderInformationEnumerator();

            foreach (BitmapCodecInformation codecInfo in codecInfos)
                foreach (string extension in codecInfo.FileExtensions)
                    picker.FileTypeFilter.Add(extension);

            // Get the selected file
            StorageFile storageFile = await picker.PickSingleFileAsync();

            if (storageFile == null)
                return;

            string exception = null;

            try
            {
                await LoadBitmapFromFile(storageFile);
            }
            catch (Exception exc)
            {
                exception = exc.Message;
            }

            if (exception != null)
            {
                MessageDialog msgdlg =
                    new MessageDialog("The image file could not be loaded. " +
                                      "The system reports an error of: " + exception,
                                      "Finger Paint");
                await msgdlg.ShowAsync();
                return;
            }

            appSettings.LoadedFilePath = storageFile.Path;
            appSettings.LoadedFilename = storageFile.Name;
            appSettings.IsImageModified = false;
        }

        async void OnSaveAppBarButtonClick(object sender, RoutedEventArgs args)
        {
            Button button = sender as Button;
            button.IsEnabled = false;

            if (appSettings.LoadedFilePath != null)
            {
                await SaveWithErrorNotification(appSettings.LoadedFilePath);
            }
            else
            {
                StorageFile storageFile = await GetFileFromSavePicker();

                if (storageFile == null)
                    return;

                await SaveWithErrorNotification(storageFile);
            }

            button.IsEnabled = true;
        }

        async void OnSaveAsAppBarButtonClick(object sender, RoutedEventArgs args)
        {
            StorageFile storageFile = await GetFileFromSavePicker();

            if (storageFile == null)
                return;

            await SaveWithErrorNotification(storageFile);
        }

        async Task<StorageFile> GetFileFromSavePicker()
        {
            FileSavePicker picker = new FileSavePicker();
            picker.SuggestedStartLocation = PickerLocationId.PicturesLibrary;
            picker.SuggestedFileName = appSettings.LoadedFilename ?? "MyFingerPainting";

            // Get the encoder information
            Dictionary<string, Guid> imageTypes = new Dictionary<string, Guid>();
            IReadOnlyList<BitmapCodecInformation> codecInfos =
                                    BitmapEncoder.GetEncoderInformationEnumerator();

            foreach (BitmapCodecInformation codecInfo in codecInfos)
            {
                List<string> extensions = new List<string>();

                foreach (string extension in codecInfo.FileExtensions)
                    extensions.Add(extension);

                string filetype = codecInfo.FriendlyName.Split(' ')[0];
                picker.FileTypeChoices.Add(filetype, extensions);

                foreach (string mimeType in codecInfo.MimeTypes)
                    imageTypes.Add(mimeType, codecInfo.CodecId);
            }

            // Get a selected StorageFile
            return await picker.PickSaveFileAsync();
        }

        async Task<bool> SaveWithErrorNotification(string filename)
        {
            StorageFile storageFile = await StorageFile.GetFileFromPathAsync(filename);
            return await SaveWithErrorNotification(storageFile);
        }

        async Task<bool> SaveWithErrorNotification(StorageFile storageFile)
        {
            string exception = null;

            try
            {
                await SaveBitmapToFile(storageFile);
            }
            catch (Exception exc)
            {
                exception = exc.Message;
            }

            if (exception != null)
            {
                MessageDialog msgdlg =
                    new MessageDialog("The image file could not be saved. " +
                                      "The system reports an error of: " + exception,
                                      "Finger Paint");
                await msgdlg.ShowAsync();
                return false;
            }

            appSettings.LoadedFilePath = storageFile.Path;
            appSettings.IsImageModified = false;
            return true;
        }

        async Task SaveBitmapToFile(string filename)
        {
            StorageFile storageFile = await StorageFile.GetFileFromPathAsync(filename);
            await SaveBitmapToFile(storageFile);
        }

        async Task SaveBitmapToFile(StorageFile storageFile)
        {
            using (IRandomAccessStream fileStream = await storageFile.OpenAsync(FileAccessMode.ReadWrite))
            {
                BitmapEncoder encoder = await BitmapEncoder.CreateAsync(BitmapEncoder.PngEncoderId, fileStream);
                encoder.SetPixelData(BitmapPixelFormat.Bgra8, BitmapAlphaMode.Premultiplied,
                                     (uint)bitmap.PixelWidth, (uint)bitmap.PixelHeight,
                                     96, 96, pixels);
                await encoder.FlushAsync();
            }
        }
    }
}
